### PR TITLE
Use AES/GCM for symmetric key wrapping

### PR DESCRIPTION
**Pull request by ellisjoe**

### CODE REVIEW

**Comment by ellisjoe** [comment]
Ran some benchmarks here: https://gist.github.palantir.build/jellis/aa80c35a565dc5b1e50de2d2f731d7c9

and found that perf wasn't a real issue with values as small as the keys we're wrapping. Think the small perf hit is worth the authentication benefits we'll get

File: crypto-keys/src/main/java/com/palantir/crypto2/keys/serialization/SymmetricKeySerializerV3.java [hunk 1]
```diff
@@ -52,9 +52,11 @@
 enum SymmetricKeySerializerV3 implements SymmetricKeySerializer {
     INSTANCE;
 
-    private static final String AES_CTR_NO_PADDING = "AES/CTR/NoPadding";
-    private static final int IV_SIZE = 16;
-
+    private static final String AES_CTR_NO_PADDING = "AES/GCM/NoPadding";
```
Comment (markelliot) [inline crypto-keys/src/main/java/com/palantir/crypto2/keys/serialization/SymmetricKeySerializerV3.java]: `AES_CTR_NO_PADDING` -> `AES_GCM_NO_PADDING`

Comment (ellisjoe) [reply crypto-keys/src/main/java/com/palantir/crypto2/keys/serialization/SymmetricKeySerializerV3.java]: done. thanks

File: crypto-keys/src/main/java/com/palantir/crypto2/keys/serialization/SymmetricKeySerializerV3.java [hunk 2]
```diff
@@ -52,9 +52,11 @@
 enum SymmetricKeySerializerV3 implements SymmetricKeySerializer {
     INSTANCE;
 
-    private static final String AES_CTR_NO_PADDING = "AES/CTR/NoPadding";
-    private static final int IV_SIZE = 16;
-
+    private static final String AES_GCM_NO_PADDING = "AES/GCM/NoPadding";
+    private static final int IV_SIZE = 12;
+    // 128 bit tag length as recommended by:
+    // https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-38d.pdf
```
Comment (uschi2000) [inline crypto-keys/src/main/java/com/palantir/crypto2/keys/serialization/SymmetricKeySerializerV3.java:58]: @cbarcenas could you take a look at this, please?

**Review by uschi2000 — APPROVED** [review]
looks good to me, but would suggest to have @cbarcenas take a peek

File: crypto-keys/src/main/java/com/palantir/crypto2/keys/serialization/SymmetricKeySerializerV3.java [hunk 3]
```diff
@@ -52,9 +52,11 @@
 enum SymmetricKeySerializerV3 implements SymmetricKeySerializer {
     INSTANCE;
 
-    private static final String AES_CTR_NO_PADDING = "AES/CTR/NoPadding";
-    private static final int IV_SIZE = 16;
-
+    private static final String AES_GCM_NO_PADDING = "AES/GCM/NoPadding";
+    private static final int IV_SIZE = 12;
```
Comment (schlosna) [inline crypto-keys/src/main/java/com/palantir/crypto2/keys/serialization/SymmetricKeySerializerV3.java:56]: Realize no-one is probably using this yet, but we're breaking compat via algorithm & IV size switch. should we embed in serialized format?

Comment (ellisjoe) [reply crypto-keys/src/main/java/com/palantir/crypto2/keys/serialization/SymmetricKeySerializerV3.java:56]: I'd personally prefer to rely on the version byte determining the algorithm and iv size. this isn't released yet which is the only reason I didn't bump the version.

**Comment by ellisjoe** [comment]
merging but will wait for @cbarcenas and a confirmation on my response to @schlosna before releasing

File: crypto-keys/src/main/java/com/palantir/crypto2/keys/serialization/SymmetricKeySerializerV3.java (hunk 2 above)
Comment (cbarcenas) [reply crypto-keys/src/main/java/com/palantir/crypto2/keys/serialization/SymmetricKeySerializerV3.java:58]: This is standard - IVs in GCM mode are usually 92 bits wide. Inside the GCM construction, a 32-bit counter suffix is appended to the input IV to form a 128-bit-wide input to the AES block function.

Comment (cbarcenas) [reply crypto-keys/src/main/java/com/palantir/crypto2/keys/serialization/SymmetricKeySerializerV3.java:58]: It's probably worth updating the inline comment to better clarify this

Comment (ellisjoe) [reply crypto-keys/src/main/java/com/palantir/crypto2/keys/serialization/SymmetricKeySerializerV3.java:58]: this comment is about the authentication tag, not the IV. I can add your comment to the IV_SIZE constant if you'd like.

**Pull request merged**